myna/database/nist_ambench_2022.py

```diff
@@ -0,0 +1,537 @@
+                )
+                times = times[:entries]
+                df = df.with_columns(pl.Series(name="Time(s)", values=times))
+
+                # Remove rows where power is zero
+                df = df.filter(pl.col("P") == 0)
+
+                # Calculate the time since the previous row (dt_last) and to the next row (dt_next)
+                active_times = df.get_column("Time(s)").to_numpy()
+                dt_last = active_times[1:] - active_times[:-1]
+                dt_last = np.insert(dt_last, 0, 0)
+                dt_next = active_times[1:] - active_times[:-1]
+                dt_next = np.append(dt_next, 0)
+                df = df.with_columns(pl.Series(name="dt_last", values=dt_last))
+                df = df.with_columns(pl.Series(name="dt_next", values=dt_next))
+
+                # Drop rows that have both dt_last and dt_next equal to SCAN_TIMESTEP
+                abs_tol = 1e-7
+                df = df.filter(
+                    ((pl.col("dt_last") - SCAN_TIMESTEP).abs() > abs_tol)
+                    | ((pl.col("dt_next") - SCAN_TIMESTEP).abs() > abs_tol)
+                )
+
+                # Add last x and y locations to each row
+                xs = df.get_column("X").to_numpy()
+                x_last = xs[:-1]
+                x_last = np.insert(x_last, 0, 0)
+                df = df.with_columns(pl.Series(name="X_last", values=x_last))
+                ys = df.get_column("Y").to_numpy()
+                y_last = ys[:-1]
+                y_last = np.insert(y_last, 0, 0)
+                df = df.with_columns(pl.Series(name="Y_last", values=y_last))
+                df = df.with_columns(
+                    (
+                        (
+                            (pl.col("X") - pl.col("X_last")).pow(2)
+                            + (pl.col("Y") - pl.col("Y_last")).pow(2)
+                        ).sqrt()
+                    ).alias("dist_to_last")
+                )
+
+                # Filter by location
+                bounds = self.get_part_bounds(part)
+                df = df.filter((pl.col("X") >= bounds[0]) & (pl.col("X") <= bounds[1]))
+                df = df.filter((pl.col("Y") >= bounds[2]) & (pl.col("Y") <= bounds[3]))
+
+                # Set up output dictionary
+                df_scan = pl.DataFrame(
+                    {
+                        "Mode": pl.Series([], dtype=pl.Int8),
+                        "X(mm)": pl.Series([], dtype=pl.Float64),
+                        "Y(mm)": pl.Series([], dtype=pl.Float64),
+                        "Z(mm)": pl.Series([], dtype=pl.Float64),
+                        "Pmod": pl.Series([], dtype=pl.Float64),
+                        "tParam": pl.Series([], dtype=pl.Float64),
+                    }
+                )
+
+                # Iterate through rows sequentially to generate scan path
+                time_last_row = None
+                for index, row in enumerate(df.iter_rows(named=True)):
```

Comment:
I think a few helper functions will clean up this loop a lot (some fields are the same in each if block, etc.)

Reply:
Added a dataframe constructor function that cleans things up a bit.